### ISSUE TITLE
findRecordsBy, findRecordsByCached, and groupBy' all reverse the order of the lists they return

### ISSUE THREAD

**Issue by liminalisht** [flipstone/orville#124]
On the `0.9_development` branch (and apparently also on `master`), `findRecordsBy` and `findRecordsByCached` do not preserve the ordering specified in any order by clauses. Instead they appear to reverse that ordering, due to a bug in `groupBy'`.

Let's say I write:
```
findRecordsBy someTable someField
$ where_
    (whereAnd
      [ ...
      ]
    )
<> O.order field1 O.Descending
<> O.order field2 O.Descending
```

I would expect this to return an `m (Map key [record])`, where the `[record]` in the `Map` is sorted in the way I defined in the order clauses, namely field1 desc, field2 desc.

What happens instead is that the order of this list is reversed.

This is because 
```
findRecordsBy ::
     (Ord fieldValue, MonadOrville conn m)
  => TableDefinition readEntity writeEntity key
  -> FieldDefinition nullability fieldValue
  -> SelectOptions
  -> m (Map.Map fieldValue [readEntity])
findRecordsBy tableDef field opts = do
  let builder = (,) <$> fieldFromSql field <*> tableFromSql tableDef
      query = selectQuery builder (fromClauseTable tableDef) opts
  Map.groupBy' id <$> runSelect query
```
and 
```
findRecordsByCached ::
     (Ord fieldValue, MonadThrow m, MonadOrville conn m)
  => TableDefinition readEntity writeEntity key
  -> FieldDefinition nullability fieldValue
  -> SelectOptions
  -> QueryCached m (Map.Map fieldValue [readEntity])
findRecordsByCached tableDef field opts = do
  let builder = (,) <$> fieldFromSql field <*> tableFromSql tableDef
  rows <- selectCachedRows tableDef opts
  Map.groupBy' id <$> unsafeLift (decodeSqlRows builder rows)
```
both rely on
```
groupBy' :: Ord k => (a -> (k, v)) -> [a] -> Map.Map k [v]
groupBy' mkEntry as = Map.fromListWith (++) (map mkListEntry as)
  where
    mkListEntry a =
      let (k, v) = mkEntry a
       in (k, [v])
```
which take your `as :: [a]`, and maps each `a` to a key/value pair, where the value is always a singleton list `[v]`. Then `Map.fromListWith (++)` is invoked on that.

Notice that this will end up reversing the elements of the list. If my `as` here are `[2, 3]`, I end up with `[3, 2]`
```
> Map.fromListWith (++) [(1,[2]),(1,[3])]                                                                                                                                                                                                                                                                                                                                                                                                                                                             
fromList [(1,[3,2])] 
```

This bug would affect anyone using `findRecordsBy`, `findRecordsByCached`, or `groupBy'` on either the `0.9_development` branch, or `master` (and other branches if they exist).

**Comment by liminalisht** [comment]
This issue presumably also affects any `Popper`s that utilize order by clauses, as `findRecordsByCached` is utilized by `popPrim`:

```
popPrim ::
     (MonadThrow m, MonadOrville conn m)
  => Prim a b
  -> a
  -> QueryCached m (Popped b)
popPrim (PrimQuery _ query) _ = PoppedValue <$> unsafeLift query
popPrim (PrimRecordBy tableDef fieldDef opts) recordId =
  PoppedValue <$>
  selectFirstCached tableDef (where_ (fieldDef .== recordId) <> opts)
popPrim (PrimRecordManyBy tableDef fieldDef opts) recordId =
  PoppedValue <$>
  selectCached tableDef (where_ (fieldDef .== recordId) <> opts)
popPrim (PrimRecordsBy tableDef fieldDef opts) recordIds =
  PoppedValue <$> Map.map head <$>
  findRecordsByCached
    tableDef
    fieldDef
    (where_ (fieldDef .<- recordIds) <> opts)
popPrim (PrimRecordsManyBy tableDef fieldDef opts) recordIds =
  PoppedValue <$>
  findRecordsByCached
    tableDef
    fieldDef
    (where_ (fieldDef .<- recordIds) <> opts)
```

**Comment by qxjit** [comment]
@liminalisht Thank you! We'll look at this and get it fixed.

**Comment by liminalisht** [comment]
I'm seeing a lot of development activity over the past 11 days, but none of the PRs coming in address this issue. It's all just on `orville-postgresql-libpq` rewrite work. I'd like to see this issue prioritized higher, and get an understanding of a timeline for a fix.

This caused a production issue for us, and is waiting to bite other teams who might use `Popper` s or `findRecordsBy` or `findRecordsByCached`. (We are still using `findRecordsBy` in production, and are temporarily re-sorting in memory, post query.)

If it's preferable, someone could grant me permissions to push a branch, and I can make the fix myself. It *looks* like a one line fix (`flip (++)` instead of `(++)`), and I would expect that adding tests should take ~30-45 min of development time.

**Comment by qxjit** [comment]
@liminalisht This is basically up next in the hopper and we should be able to get to it by the end of week. Sorry it's taken longer than expected. It's been hard for us to switch context from things we had in flight already with what time we have to dedicate to Orville at the moment.

**Comment by telser** [comment]
This should be fixed in d0c3c4733cf6b04da5c1c73ab94aa2e7af0361ae which is PR #128 from above. I'm going to close out this issue, but please re-open (or file a new issue) if we missed something in the code and tests meant to address this.
